jmespath/lexer.py

```diff
@@ -6,143 +6,170 @@
+                yield {'type': 'number', 'value': int(buff),
+                       'start': start, 'end': start + len(buff)}
+            elif self._current == '"':
+                yield self._consume_quoted_identifier()
+            elif self._current == '<':
+                yield self._match_or_else('=', 'lte', 'lt')
+            elif self._current == '>':
+                yield self._match_or_else('=', 'gte', 'gt')
+            elif self._current == '!':
+                yield self._match_or_else('=', 'ne', 'unknown')
+            elif self._current == '=':
+                yield self._match_or_else('=', 'eq', 'unknown')
+            else:
+                raise LexerError(lexer_position=self._pos,
+                                 lexer_value=self._current,
+                                 message="Unknown token %s" % self._current)
+        yield {'type': 'eof', 'value': '',
+               'start': self._len, 'end': self._len}
+
+    def _init_expr(self, expression):
         if not expression:
             raise EmptyExpressionError()
-        previous_column = 0
-        for match in self.master_regex.finditer(expression):
-            value = match.group()
-            start = match.start()
-            end = match.end()
-            if match.lastgroup == 'skip':
-                # Ignore whitespace.
-                previous_column = end
-                continue
-            if start != previous_column:
-                bad_value = expression[previous_column:start]
-                # Try to give a good error message.
-                if bad_value == '"':
-                    raise LexerError(
-                        lexer_position=previous_column,
-                        lexer_value=value,
-                        message='Starting quote is missing the ending quote',
-                        expression=expression)
-                raise LexerError(lexer_position=previous_column,
-                                 lexer_value=value,
-                                 message='Unknown character',
-                                 expression=expression)
-            previous_column = end
-            token_type = match.lastgroup
-            handler = getattr(self, '_token_%s' % token_type.lower(), None)
-            if handler is not None:
-                value = handler(value, start, end)
-            yield {'type': token_type, 'value': value,
-                   'start': start, 'end': end}
-        # At the end of the loop make sure we've consumed all the input.
-        # If we haven't then we have unidentified characters.
-        if end != len(expression):
-            msg = "Unknown characters at the end of the expression"
-            raise LexerError(lexer_position=end,
-                             lexer_value='',
-                             message=msg, expression=expression)
+        self._pos = 0
+        self._expression = expression
+        self._chars = list(self._expression)
```

Comment:
Huh, that's counterintuitive.  I would not have expected that to make things faster but it does:

```
$ python -m timeit -s 'mylist=list("asdfasfasfasfasfasdfasdfasfasdabcdefgasdfasdfasdfasdfasdfasfasdf")' 'mylist[60]'
10000000 loops, best of 3: 0.0308 usec per loop

$ python -m timeit -s 'mystr="asdfasfasfasfasfasdfasdfasfasdabcdefgasdfasdfasdfasdfasdfasfasdf"' 'mystr[60]'
10000000 loops, best of 3: 0.0396 usec per loop
```

Good to know.